[BE] document more of FSDP checkpointing logic with a sprinkle of cleaning

janeyx99 opened the pull request:
This PR should not make any functional difference. It:
- adds clearer documentation
- clarifies a type
- revises minor typos
- swaps a .keys for a .items call on a dictionary

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106069

janeyx99 committed:
[BE] document more of FSDP checkpointing logic with a sprinkle of cleaning

[ghstack-poisoned]

janeyx99 committed:
Update on "[BE] document more of FSDP checkpointing logic with a sprinkle of cleaning"

This PR should not make any functional difference. It:
- adds clearer documentation
- clarifies a type
- revises minor typos
- swaps a .keys for a .items call on a dictionary




[ghstack-poisoned]

fegin commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -155,8 +155,7 @@ def _unflatten_optim_state(
         for optim_state in unflat_param_state:
-            for key in list(optim_state.keys()):
-                state = optim_state[key]
+            for key, state in optim_state.items():
```
We cannot do this. Python will complain optim_state is being modified while looping.

janeyx99 replied:
Oh I see. I'll leave a comment instead

fegin reviewed (commented):
Thanks for the BE effort. There is one change will break the code. Please see the comment.

janeyx99 committed:
Update on "[BE] document more of FSDP checkpointing logic with a sprinkle of cleaning"

This PR should not make any functional difference. It:
- adds clearer documentation
- clarifies a type
- revises minor typos
- swaps a .keys for a .items call on a dictionary




[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -266,11 +266,23 @@ def _get_param_to_fqns(
+    Constructs a mapping from parameter to a list of its \"canonical\" FQNs. Here,
+    we use canonical to mean the fully-qualified name assigned to the parameter
+    based on its position in the nn.Module hierarchy. This is in contrast to FQNs
+    that may be generated by other parallelisms or wrappers.
```
"generated" --> does this mean `named_parameters()`?

janeyx99 replied:
yes, but i'm moreso referring to the fact that FlatParams get swizzled into the modules and have their own names

awgu commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -266,11 +266,23 @@ def _get_param_to_fqns(
+    have length greater than one.  All FQNs are prefixed starting from ``model``.
+
+    In the case where FSDP was applied with ``use_orig_params=True``, there should be no
+    ``FlatParameter`` s in the model and this mapping will only contain mappings from
```
```suggestion
    ``FlatParameter`` s registered to the model's modules and this mapping will only contain mappings from
```

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -990,7 +991,20 @@ def _get_param_id_to_param_from_optim_input(
+    qualified name (FQN) as opposed to the FQNs returned for ``FlatParameter`` s
+    in ``_common_utils._get_param_to_fqns(...)``. We refer to these FQNs as
+    non-canonical because these parameters are not from the original module
+    but are rather swapped in by some parallelism construct, like FSDP.
```
I do not follow what "swapped in by some parallelism construct, like FSDP" means.

The distinction seems to be that FQNs here have been cleaned via `clean_tensor_name()` to remove the FSDP prefix. Is this what you mean by "swap"?

janeyx99 replied:
My understanding is that when you do use_orig_params=False, FSDP swaps out original module parameters with a FlatParameter that represents the original parameters.

This new FlatParameter will then have its own name that FSDP gave, and the module will use that instead of the original FQNs of the original params that got swapped out.

Maybe my understanding's not right here though--would be happy to rephrase

janeyx99 committed:
Update on "[BE] document more of FSDP checkpointing logic with a sprinkle of cleaning"

This PR should not make any functional difference. It:
- adds clearer documentation
- clarifies a type
- revises minor typos
- swaps a .keys for a .items call on a dictionary




[ghstack-poisoned]

awgu reviewed (approved):
LGTM!

awgu commented:
@pytorchbot merge
